examples/pull.rs

```diff
@@ -0,0 +1,200 @@
+#[macro_use]
+extern crate serde_derive;
+
+use docopt::Docopt;
+use git2::Repository;
+use std::io::{self, Write};
+use std::str;
+
+#[derive(Deserialize)]
+struct Args {
+    arg_remote: Option<String>,
+    arg_branch: Option<String>,
+}
+
+fn do_fetch<'a>(
+    repo: &'a git2::Repository,
+    refs: &[&str],
+    remote: &'a mut git2::Remote,
+) -> Result<git2::AnnotatedCommit<'a>, git2::Error> {
+    let mut cb = git2::RemoteCallbacks::new();
+
+    // Print out our transfer progress.
+    cb.transfer_progress(|stats| {
+        if stats.received_objects() == stats.total_objects() {
+            print!(
+                "Resolving deltas {}/{}\r",
+                stats.indexed_deltas(),
+                stats.total_deltas()
+            );
+        } else if stats.total_objects() > 0 {
+            print!(
+                "Received {}/{} objects ({}) in {} bytes\r",
+                stats.received_objects(),
+                stats.total_objects(),
+                stats.indexed_objects(),
+                stats.received_bytes()
+            );
+        }
+        io::stdout().flush().unwrap();
+        true
+    });
+
+    let mut fo = git2::FetchOptions::new();
+    fo.remote_callbacks(cb);
+    fo.download_tags(git2::AutotagOption::All); // Always fetch all tags.
+                                                // Perform a download and also update tips
+    println!("Fetching {} for repo", remote.name().unwrap());
+    remote.fetch(refs, Some(&mut fo), None)?;
+
+    {
+        // If there are local objects (we got a thin pack), then tell the user
+        // how many objects we saved from having to cross the network.
+        let stats = remote.stats();
+        if stats.local_objects() > 0 {
+            println!(
+                "\rReceived {}/{} objects in {} bytes (used {} local \
+                 objects)",
+                stats.indexed_objects(),
+                stats.total_objects(),
+                stats.received_bytes(),
+                stats.local_objects()
+            );
+        } else {
+            println!(
+                "\rReceived {}/{} objects in {} bytes",
+                stats.indexed_objects(),
+                stats.total_objects(),
+                stats.received_bytes()
+            );
+        }
+    }
+
+    let fetch_head = repo.find_reference("FETCH_HEAD")?;
+    Ok(repo.reference_to_annotated_commit(&fetch_head)?)
+}
+
+fn fast_forward(lb: &mut git2::Reference, rc: &git2::AnnotatedCommit) -> Result<(), git2::Error> {
+    let name = match lb.name() {
+        Some(s) => s.to_string(),
+        None => String::from_utf8_lossy(lb.name_bytes()).to_string(),
+    };
+    let msg = format!("Fast-Forward: Setting {} to id: {}", name, rc.id());
+    lb.set_target(rc.id(), &msg)?;
+    Ok(())
+}
+
+fn normal_merge(
+    repo: &Repository,
+    local: &git2::AnnotatedCommit,
+    remote: &git2::AnnotatedCommit,
+) -> Result<(), git2::Error> {
+    let local_tree = repo.find_tree(local.id())?;
```

Comment:
This code path isn't working for me.  I am getting:

`error: the requested type does not match the type in ODB; class=Invalid (3); code=NotFound (-3)`

Since `local` is a commit, I assume `find_tree` does not dereference the commit to a tree.  I'm not sure how this should work.

Reply:
Hrmmm. Your right that does indeed fail. I'm not sure what the proper way to do this is actually. Looks like I have some reading to do.

Reply:
Aha I was getting mired in the whole object db semantics. looks like the correct way is to do
`let local_tree = repo.find_commit(local.id())?.tree()?'`. I'll try to test this a little bit later when I have some more time this weekend.

Reply:
@ehuss I've tested this and it seems to work. I've also handled the conflicts on merge case.